src/reflect/scala/reflect/internal/tpe/TypeComparers.scala

```diff
@@ -351,6 +351,7 @@ trait TypeComparers {
   // @assume tp1.isHigherKinded || tp2.isHigherKinded
   def isHKSubType(tp1: Type, tp2: Type, depth: Depth): Boolean = {
     def isSub(ntp1: Type, ntp2: Type) = (ntp1.withoutAnnotations, ntp2.withoutAnnotations) match {
+      case (TypeRef(_, AnyClass, _), PolyType(_, _))                        => true                     // Any is kind-polymorphic
```

Comment:
As far as I can tell, the main problem to solve is over in `instantiateTypeVars` and `setInst`, which temporarily mutate the info of the enclosing type parameter while typechecking GADT refined cases. They end up assigning a `TypeBounds(lo, hi)` to the type param, which is never seen as being higher-kinded itself (even if `lo` and `hi` are). That ends up tripping up the kind conformance check.

I'd suggest trying something like:

https://github.com/scala/scala/compare/2.12.x...retronym:ticket/10208?expand=1